libtransmission/torrent.c

```diff
@@ -1131,6 +1132,72 @@ static tr_parse_result torrentParseImpl(tr_ctor const* ctor, tr_info* setmeInfo,
     return result;
 }
 
+static void tr_torrentAddDefaultTrackers(tr_torrent* tor)
+{
+    tr_list* l;
+    int tier = -1;
+    int i;
+    tr_tracker_info* trackers;
+    int numExistingTrackers = tor->info.trackerCount;
+    int numNewTrackers = tr_list_size(tor->session->defaultTrackers);
+    char* url;
+    bool duplicate;
+    bool changed = false;
+
+    if (!numNewTrackers || tor->info.isPrivate)
+    {
+        return;
+    }
+
+    // copy existing trackers
+    trackers = tr_new0(tr_tracker_info, numExistingTrackers + numNewTrackers);
+    for (i=0; i<numExistingTrackers; ++i)
+    {
+        trackers[i].tier = tor->info.trackers[i].tier;
+        trackers[i].announce = tr_strdup(tor->info.trackers[i].announce);
+        tier = MAX(tier, tor->info.trackers[i].tier);
+    }
+
+    // add the new ones
+    for (l=tor->session->defaultTrackers; l!=NULL; l=l->next)
+    {
+        url = (char*)l->data;
+        if (tr_urlIsValidTracker(url))
+        {
+            // check for duplicates
+            duplicate = false;
+            for (i=0; i<numExistingTrackers; ++i)
+            {
+                if (strcmp(trackers[i].announce, url) == 0)
+                {
+                    duplicate = true;
+                    break;
+                }
+            }
+            if (duplicate)
+            {
+                continue;
+            }
```

Comment:
I feel like this block duplicates the helper function `findAnnounceUrl()`. Maybe that utility could be moved from `rpcimpl.c` to `torrent.[ch]` so that it could be reused here?

Reply:
Yes, but the function name is misleading in this context, rendering the code less readable.